### ISSUE TITLE
Support component inheritance

### ISSUE THREAD

**Issue by f-mon** [dart-lang/angular#231]
At current time Angular Dart doesn't support component hineritance of @component and @input  and other metadata. This very important feature has been added in angular 2.3.0, are there any plan to support this in angular dart?

**Comment by daniel-v** [comment]
Might I ask what is this blocked by? This would be a neat feature.

**Comment by TedSander** [comment]
We need to use resolved ASTs for codegen to enable this. The blocker is currently being worked on.

**Comment by matanlurey** [comment]
@f-mon @daniel-v This will be in 3.0.0, though I can't say if it will be immediate or not.

**Comment by matanlurey** [comment]
Pushed to 4.0.

**Comment by matanlurey** [comment]
This is now planned for the 4.0 release.

**Comment by matanlurey** [comment]
If anyone would like to start an RFC on how this should work (or refer to TypeScript's implementation) that would be awesome. There is a chance this might slip from `4.0` but I'd personally love to see it.

**Comment by Lisenish** [comment]
I would like to participate in this.
Also, I want to note it would be great to implement **@component** annotation inheritance not only @Input like on TS Angular, as we can see in this issue https://github.com/angular/angular/issues/7968 this is a highly requested feature but looks like guys from TS version doesn't agree with the community proposal. So it can be the small killer feature of Angular Dart vs TS Angular =).

**Comment by leonsenft** [comment]
We plan to support inheritance for all property annotations in 4.0.0.

- `@ContentChild`
- `@ContentChildren`
- `@HostBinding`
- `@HostListener`
- `@Input`
- `@Output`
- `@ViewChild`
- `@ViewChildren`

Derived components will inherit their ancestor's property annotations, and may choose to override their implementations. If a derived component re-declares a named annotation on a different property, we'll treat this similarly to how duplicate annotations in a component currently behave.

We won't be adding support for inheritance of class annotations yet.

- `@Component`
- `@Directive`
- `@Pipe`

**Comment by leonsenft** [comment]
Thank you for your support and interest @Lisenish. If you have any feedback on our current plans I'd be glad to hear it. If you feel strongly about inheritable class annotations, please feel to provide some details on how you'd like that to work.

There's not one clear way to implement class annotations. Since we haven't seen much demand for it yet, there's not much sense in committing to one arbitrary solution

**Comment by zoechi** [comment]
> We won't be adding support for inheritance of class annotations yet.

I assume this just means annotations of sub and superclass aren't "merged", but need to repeated with all parameters on the subclass, right?

**Comment by matanlurey** [comment]
It means the top-level `@Component|Directive|Pipe` is not inherited, only individual annotations on fields. For example, this would be inherited:

```dart
class AComp {
  @Input()
  String a;
}

// Still has an @Input() for a.
class BComp extends AComp {}
```

**Comment by Lisenish** [comment]
@leonsenft Good to hear, thanks! 
About annotation inheritance, this is mostly about reusing parent component template and providers. I have the case where I need child component that differs from the parent only in a selector and some business logic. Now I have to duplicate component template in child component
I understand that class annotation inheritance isn't very transparent and why you are trying to avoid this feature. But maybe we can figure out how to deal with this duplication?

**Comment by leonsenft** [comment]
@Lisenish 
If your desire is to inherit the template and providers as-is, you can already do so without duplication by using `templateUrl` and a predefined list of providers. 

```dart

const baseProviders = const [...];

@Component(
  selector: 'base',
  templateUrl: 'base_component.html',
  providers: baseProviders,
)
class BaseComponent {
  ...
}

@Component(
  selector: 'derived',
  templateUrl: 'base_component.html', // reuse supertype template URL
  providers: baseProviders, // reuse supertype providers
)
class DerivedComponent extends BaseComponent {
  ...
}
``` 
Yes, you still have to specify the `templateUrl` and `providers` argument in the component annotation, but this is a tiny, constant amount of work per component.

My concern with your proposed solution is it's all or nothing. You either inherit the base metadata as-is, or replace it entirely. A solution where the derived component can interact in some way with its inherited metadata is much more compelling.

Notice that all of the metadata we're initially supporting annotates a component method or property. Methods and property accessors can reference their inherited implementations via `super` and extend on them if desired.

No parallel to this concept exists for annotation arguments. For example, a `@Component` annotation on a derived component could not reference the template specified in the `@Component` annotation of the base component.

Just to be clear, we're not ruling out any particular implementation yet. We want to be sure that if and when we implement this, we get it right.

I'd like to reiterate my earlier point that we just haven't seen much of a demand for inheritable templates/styles/providers and how particular implementations could solve problems in ways the framework currently struggles to. If you have a compelling example, we'd be happy to see a new issue with a detailed proposal.

**Comment by zoechi** [comment]
Polymer dropped template inheritance before 1.0 where the subclass can extend the template, because it was too complicated.

**Comment by Lisenish** [comment]
@leonsenft Thanks for the detailed answer and clarification! 
Yeah, I was thinking about that solution that you proposed too after I had wrote comment. 
Actually, I changed my mind I would agree with you, looks like it's pretty difficult to implement and say how it should work while profit is minimal.
For now, we are waiting Input and View/ContentChild inheritance, this is must have for sure!

**Comment by matanlurey** [comment]
@leonsenft your latest commits landed into master - do you want to mark this done?

**Comment by leonsenft** [comment]
@matanlurey Not yet; only very limited support is landed. Currently components only support inheritance from ancestor components and directives. I have a pending change which extends inheritance to all ancestors, so you can annotate regular classes with metadata.
